cpp/dolfinx/fem/DofMap.cpp

```diff
@@ -157,6 +157,54 @@ fem::DofMap build_collapsed_dofmap(MPI_Comm comm, const DofMap& dofmap_view,
 }
 } // namespace
 
+//-----------------------------------------------------------------------------
+graph::AdjacencyList<std::int32_t>
+fem::transpose_dofmap(graph::AdjacencyList<std::int32_t>& dofmap,
+                      std::int32_t num_cells)
+{
+  // Count number of cell contributions to each global index
+  const std::int32_t max_index
+      = dofmap.array().head(dofmap.offsets()(num_cells)).maxCoeff();
+  std::vector<int> dofs_per_cell(num_cells);
+  std::vector<int> num_local_contributions(max_index + 1);
+  for (int c = 0; c < num_cells; ++c)
+  {
+    auto dofs = dofmap.links(c);
+    dofs_per_cell[c] = dofs.size();
+
+    // Count cells that contribute to a global dof index
+    for (int i = 0; i < dofs.rows(); ++i)
+      num_local_contributions[dofs[i]]++;
+  }
+
+  // Compute offset for each global index
+  std::vector<int> index_offsets(num_local_contributions.size() + 1, 0);
+  std::partial_sum(num_local_contributions.begin(),
+                   num_local_contributions.end(), index_offsets.begin() + 1);
+
+  // Compute offset for each cell block
+  std::vector<int> cell_offsets(dofs_per_cell.size() + 1, 0);
+  std::partial_sum(dofs_per_cell.begin(), dofs_per_cell.end(),
+                   cell_offsets.begin() + 1);
```

Comment:
This can be simplified by calculating the cell offsets on the fly in the next loop, eliminating both `cell_offsets` and `dofs_per_cell`.

Reply:
@chrisrichardson can you make the change in this branch?